docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,401 @@
+// specific block that will not necessarily be included in the block itself.
+// This data is obtained from the FinalizeBlock ABCI response.
+message ExecBlockResults {
+    repeated Event events = 1;
+    repeated ExecTxResult tx_results = 2;
+}
+
+// CommitBlockResponse is either empty or returns an error. Note that returning
+// an error here will cause Tendermint to crash.
+message CommitBlockResponse {
+    oneof error {
+        UnexpectedBlockError unexpected_block_err = 1;
+    }
+}
+
+// UnexpectedBlockError is an error returned by the server when Tendermint sent
+// it a block that is ahead of the block expected by the server.
+message UnexpectedBlockError {
+    // The height of the block expected by the server.
+    int64 expected_height = 1;
+}
+
+//-----------------------------------------------------------------------------
+// The following types are defined as part of ABCI++, and provided here for
+// reference.
+//-----------------------------------------------------------------------------
+
+// Event allows application developers to attach additional information to
+// ResponseFinalizeBlock, ResponseDeliverTx, ExecTxResult
+// Later, transactions may be queried using these events.
+message Event {
+  string                  type       = 1;
+  repeated EventAttribute attributes = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "attributes,omitempty"];
+}
+
+// EventAttribute is a single key-value pair, associated with an event.
+message EventAttribute {
+  string key   = 1;
+  string value = 2;
+  bool   index = 3;  // nondeterministic
+}
+
+// ExecTxResult contains results of executing one individual transaction.
+//
+// * Its structure is equivalent to #ResponseDeliverTx which will be deprecated/deleted
+message ExecTxResult {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5;
+  int64          gas_used   = 6;
+  repeated Event events     = 7
+      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];  // nondeterministic
+  string codespace = 8;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should allow for configurable buffering of a reasonable, yet small,
+number of requests to the companion service. If this buffer fills up, we should
+panic.
+
+This data would need to be stored on disk to ensure that Tendermint could
+attempt to resubmit all unsubmitted data upon Tendermint startup. At present,
+this data is already stored on disk, and so practically we would need to
+implement some form of background pruning mechanism to remove the data we know
+has been shared with the companion service.
+
+### Interaction with Block Sync
+
+During block sync, request buffering should be disabled and requests to the
+companion service should be blocking.
```

Comment:
I think we should consider have a buffer of 1 just so the block sync protocol is capable of moving on to fetching the next block, while the data service processes the data. I'm assuming the calls are synchronous as the server responds with an error if the operation failed in some way

Reply:
Just to be explicit about syncing between the data service and the consensus node. If the data service were somehow to be ahead of the node, it should just skip processing until a request with the next height is sent. If the node is ahead of the data service I'm guessing the data service will error and the node operator would need to roll back the consensus node to the same height (is that right?) or would we persist some amount of responses to disk so we could replay it to the data service?